third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -21,15 +21,178 @@ namespace nvfuser {
 
 namespace {
 
-// round up to multiple of 8 or pow2 whichever smaller
-int64_t roundUpPow2Or8(const int64_t x) {
-  auto round_up_pow2 = scheduler_utils::lastPow2(x);
-  if (round_up_pow2 < x) {
-    round_up_pow2 *= 2;
+std::shared_ptr<ReductionParams> innerOuterPersistentHeuristic(
+    const int64_t total_reduction_numel,
+    const int64_t total_iteration_numel,
+    const int64_t inner_most_dimension_numel,
+    const int64_t n_tensor_inputs,
+    const int64_t max_input_dtype_size,
+    const int64_t max_persistent_buffer_size,
+    const size_t vectorize_factor) {
+  auto rparams = std::make_shared<ReductionParams>();
+  rparams->persistent_kernel = true;
+  rparams->fastest_dim = true;
+  rparams->combined_inner_outer = true;
+  if (inner_most_dimension_numel <= 1024) {
+    rparams->multiple_reds_per_blk = true;
+  }
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t device_multiprocessor_count =
+      (int64_t)dev_prop->multiProcessorCount;
+
+  auto getRegPerThread = [&](const int threadsPerBlock, const int blocksPerSM) {
+    int64_t warpsPerSM =
+        ceilDiv(threadsPerBlock, dev_prop->warpSize) * blocksPerSM;
+    constexpr int reg_allocation_granularity = 256;
```

Comment:
Let's use cudaOccSubPartitionsPerMultiprocessor and    cudaOccRegAllocationGranularity. See executor_utils.cpp. Seems like this should be refactored as a util function.

Reply:
refactored to `getRegPerThreadGivenThreadsPerSM`